### PR TITLE
Beam search

### PR DESCRIPTION
**Pull request by mberaha**

### CODE REVIEW

File: beam_search_testing.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
+from src.parameters import Params
```
Comment (mberaha) [inline beam_search_testing.py]: rm

File: scripts/train_yelp.py [hunk 2]
```diff
@@ -19,8 +18,14 @@
 
     # unCOMMENT this if running from environments like Notebooks, Hydrogen...
     args = easydict.EasyDict({
-            "train_files": "data/yelp/train/*",
-            "evaluation_files": "data/yelp/dev/*",
+            "train_file_positive": "data/yelp/train/positive_sentence.txt",
```
Comment (mberaha) [inline scripts/train_yelp.py]: rm

File: scripts/train_yelp.py [hunk 3]
```diff
@@ -33,8 +38,21 @@
...
+        batchsize=params.batch_size,
+        inMemory=True)
+
+    validSet = batchesFromFiles(
+        positiveFile=args.evaluation_file_positive,
+        negativeFile=args.evaluation_file_negative,
+        batchsize=params.batch_size,
+        inMemory=True)
+
+    # print(trainBatches)
```
Comment (mberaha) [inline scripts/train_yelp.py]: rm

File: src/base_model.py [hunk 4]
```diff
@@ -10,8 +10,8 @@ def __init__(self):
         super().__init__()
         self.iter = 0
 
-    def load(self):
-        self.load_state_dict(self.params.savefile)
+    def load(self, fileName):
```
Comment (mberaha) [inline src/base_model.py:13]: bug!!!

File: src/beam_search.py [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+            word -- the id of the word charachterising the state
+            h -- the hidden state associated to that state
+            sentence -- a list of word ids (the past ids plus the current one)
+            nll -- the negative log likelihood corresponding to the sentence
+        """
+        self.word, self.h, self.sentence, self.nll = \
+            word, h, sentence, nll
+
+
+class Decoder(object):
```
Comment (mberaha) [inline src/beam_search.py]: BeamSearchDecoder

File: src/beam_search.py [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+        Args:
+            tokens --
+            h --
+        Outputs:
+            logProbs --
+            indices --
+            h --
+        """
+        # embedding nn lookup
+        # currTokens = torch.matmul(
```
Comment (mberaha) [inline src/beam_search.py]: rm

File: src/beam_search.py [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+        # smooth logits into the probabilities of each word
+        vocabProbs = F.softmax(
+            vocabLogits / self.params.temperature, dim=1)
+        # beam search trick to prevent probs vanishing
+        logProbs = torch.log(vocabProbs)
+        # take the beam_with most probable words
+        logProbs, indices = torch.topk(logProbs, self.width, dim=-1)
+        return logProbs, indices, h
+
+    def printDebug(self, storeBeamLayer):
```
Comment (mberaha) [inline src/beam_search.py]: rm

File: src/beam_search.py [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+        init_state = BeamState(
+            go,
+            h0,
+            [[] for i in range(batch_size)],
+            [0]*batch_size)
+        beam = [init_state]
+        for _ in range(self.max_length):
+            storeBeamLayer = [[] for _ in range(batch_size)]
+            for state in beam:
+                # self.printDebug(storeBeamLayer)
```
Comment (mberaha) [inline src/beam_search.py]: rm

File: src/greedy_decoding.py [hunk 9]
```diff
@@ -0,0 +1,66 @@
+import torch
+import torch.nn as nn
+
+device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+
+
+class Decoder(object):
```
Comment (mberaha) [inline src/greedy_decoding.py]: GreedyDecoder

File: src/greedy_decoding.py [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+            h0s, self.max_len, evaluation=True, soft=False)
+        sentences = []
+        for sentIndex in range(batch_size):
+            sent = tokens[sentIndex, :]
+            sentences.append(
+                [self.model.vocabulary.id2word[int(sent[i])]
+                 for i in range(self.max_len)])
+        return sentences
+
+    def _decode(self, h0s):
```
Comment (mberaha) [inline src/greedy_decoding.py]: da notebook

File: src/parameters.py [hunk 11]
```diff
@@ -22,16 +22,20 @@ def __init__(self, embedding_size, dim_y, dim_z):
...
     embedding_size = 200
     dim_y = 200
     dim_z = 500
-    batch_size = 128
-    epochs = 10  # 2 for debugging, then go to 10
-    temperature = 0.001
+    batch_size = 256
+    epochs = 20  # 2 for debugging, then go to 10
+    temperature = 0.1  # softmax(logit / gamma)
+    # gamma_decay = 1
```
Comment (mberaha) [inline src/parameters.py]: rm

File: src/style_transfer.py [hunk 12]
```diff
@@ -35,7 +36,10 @@ def __init__(self, params, vocabulary: Vocabulary):
             batch_first=True).to(device)
 
         # instantiating linear networks for hidden transformations
-        self.labelsTransform = torch.nn.Linear(1, params.dim_y).to(device)
+        self.encoderLabelsTransform = \
```
Comment (mberaha) [inline src/style_transfer.py:39]: @WolfLo faceva così anche shen? cioè avere due transformer dei labels?

File: src/style_transfer.py [hunk 13]
```diff
@@ -82,259 +87,308 @@ def __init__(self, params, vocabulary: Vocabulary):
...
-        loss_d = self.adv_loss_criterion(class_real, labels) + \
-            self.adv_loss_criterion(class_fake, 1 - labels)
-        loss_g = self.adv_loss_criterion(class_fake, ones)
-        return loss_d, loss_g
-
-    def _zeroGradients(self):
-        self.autoencoder_optimizer.zero_grad()
-        self.discriminator0_optimizer.zero_grad()
-        self.discriminator1_optimizer.zero_grad()
+        # tokens = torch.cat((goEmbedding, tokens), dim=1)
```
Comment (mberaha) [inline src/style_transfer.py]: rm

File: src/style_transfer.py [hunk 14]
```diff
@@ -82,259 +87,308 @@ def __init__(self, params, vocabulary: Vocabulary):
...
         # generating the hidden states with inverted labels (yq, zp)
-        transformedHiddens = self.labelsTransform(1 - tensorLabels)
+        transformedHiddens = self.generatorLabelsTransform(1 - tensorLabels)
         transformedHiddens = transformedHiddens.unsqueeze(0)
-        transformedHiddens = torch.cat(
+        self.transformedHiddens = torch.cat(
             (transformedHiddens, content), dim=2)
 
-        # reconstruction loss
+        # print("content.shape", content.shape)
```
Comment (mberaha) [inline src/style_transfer.py]: rm

File: src/style_transfer.py [hunk 15]
```diff
@@ -82,259 +87,308 @@ def __init__(self, params, vocabulary: Vocabulary):
...
+        which_params - string 'd0' or 'd1' or 'eg'
+        """
+
+        self.losses = defaultdict(float)
+
+        negativeIndex = [
+            i for i in range(self.params.batch_size//2)]
+        positiveIndex = [
+            i for i in range(self.params.batch_size//2, self.params.batch_size)]
+        # negativeIndex = np.where(labels == 0)[0]
```
Comment (mberaha) [inline src/style_transfer.py]: meno generale ma va bene lo stesso!

File: src/style_transfer.py [hunk 16]
```diff
@@ -82,259 +87,308 @@ def __init__(self, params, vocabulary: Vocabulary):
...
+
+        return encoder_inputs, generator_inputs, targets, lengths
 
     def trainOnBatch(self, sentences, labels, iterNum):
         self.train()
         labels = np.array(labels)
         encoder_inputs, generator_inputs, targets, lenghts = \
             self._sentencesToInputs(sentences)
+
+        # print("trainOnBatch encoder_inputs.shape:", encoder_inputs.shape)
```
Comment (mberaha) [inline src/style_transfer.py]: rm

File: src/tests.py [hunk 17]
```diff
@@ -0,0 +1,144 @@
+import torch
```
Comment (mberaha) [inline src/tests.py]: rm

File: src/style_transfer.py (hunk 15 above)
Comment (mberaha) [reply src/style_transfer.py]: in realtà ho cambiato...
